codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -121,35 +123,84 @@ public void close() { }
     @Override
     public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data,
             int padding, boolean endStream, ChannelPromise promise) {
-        boolean releaseData = true;
+        boolean lastHeaderWritten = false;
+        boolean lastFrameWritten = false;
         SimpleChannelPromiseAggregator promiseAggregator =
                 new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+
+        // All padding will be put at the end, so in the worst case we need 3 headers:
+        // a repeated no-padding frame of maxFrameSize, a frame that has part data and part
+        // padding, and a frame that has the remainder of the padding.
+        ByteBuf headerBuf = ctx.alloc().buffer(3 * DATA_FRAME_HEADER_LENGTH);
         try {
             verifyStreamId(streamId, STREAM_ID);
             verifyPadding(padding);
 
-            Http2Flags flags = new Http2Flags().paddingPresent(padding > 0).endOfStream(endStream);
+            int remainingData = data.readableBytes();
+            int headerIndex = 0;
+            int prevFrameDataBytes = 0;
+            int prevFramePaddingBytes = 0;
+            ByteBuf frameHeader = null;
 
-            int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
-            verifyPayloadLength(payloadLength);
+            do {
+                // Determine how much data and padding to write in this frame.
+                int frameDataBytes = min(remainingData, maxFrameSize);
+                int framePaddingBytes = min(padding, max(0, (maxFrameSize - 1) - frameDataBytes));
+
+                // Decrement the remaining counters.
+                padding -= framePaddingBytes;
```

Comment:
given that the job of padding is to obfuscate payload length I think that placing it at the end is perfectly reasonable and probably more efficient so I have no issue with what this code does. 

The code could be written to always put padding into it's own frame entirely. This would simplify the code a little but waste a small amount of bandwidth. Not an unreasonable tradeoff IMHO but not one Im going to push for either.